### ISSUE TITLE
`task run --generate-cmd` gives invalid output if an env var has commas in it

### ISSUE THREAD

**Issue by codekitchen** [aws/copilot-cli#3683]
Example of the error:

```bash
❯ copilot task run --env-vars a="1,2,3",b="b"
✘ invalid argument "a=1,2,3,b=b" for "--env-vars" flag: 2 must be formatted as key=value
```

As far as I've been able to determine, the error message comes from the [pflag library](https://pkg.go.dev/github.com/spf13/pflag@v1.0.5#section-readme) but this seems to actually be a bug in `--generate-cmd`... the quotes that `--generate-cmd` wraps the values in get stripped out by the shell, so their effect is to shell-quote but not to delimit env vars to copilot. You can see this more simply by running echo:

```bash
❯ echo --env-vars a="1,2,3",b="b"
--env-vars a=1,2,3,b=b
```

I tried escaping the quotes, thinking maybe copilot could start escaping them, but that fails too:

```bash
❯ copilot task run --env-vars a=\"1,2,3\",b=\"b\"       
✘ invalid argument "a=\"1,2,3\",b=\"b\"" for "--env-vars" flag: parse error on line 1, column 3: bare " in non-quoted-field
```

What _does_ work is passing each env var as its own arg, rather than one arg separated by commas:

```bash
❯ copilot task run --env-vars "a=1,2,3" --env-vars "b=b"
```

I'm not sure what the preferred fix would be here, though. Thoughts?

**Comment by iamhopaul123** [comment]
Hello @codekitchen. <strike>It seems like a limitation for the pflag package. In this scenario I think using flag repetition will be the only recommended way if the value contains any comma (related to https://github.com/spf13/pflag/issues/276) 😕</strike>

Actually this works for me

```
copi task run --env-vars \"a=1,2,3\",\"b=asdf\"
```

**Comment by codekitchen** [comment]
ah nice I didn't try escaped-quoting both the key and value together, that makes sense.

which fix would the copilot team prefer for the output of `--generate-cmd`? I'm happy to send over a PR for either one.

**Comment by iamhopaul123** [comment]
Right right it was originally because of https://github.com/aws/copilot-cli/issues/3458 and i'm testing see if the new format works for secrets, which in our current mindset require the value to be quoted.

**Comment by iamhopaul123** [comment]
OK it seems like it works for secrets. We might just need to update the quote to make sure it covers both key and value together. Something like this
```
copi task run
--execution-role arn:aws:iam::1234567890:role/demo-test-frontend-ExecutionRole-VFSF440B
J7VV 
--task-role arn:aws:iam::1234567890:role/demo-test-frontend-TaskRole-1FAN90CIBJFIU
--image 1234567890.dkr.ecr.us-west-2.amazonaws.com/demo/frontend@sha256:50f
1329d876c9c81b363503cdbedc7ed57e77e6301c34346e8aa0d603c9ab469
--entrypoint ""
--command ""
--env-vars \"COPILOT_APPLICATION_NAME=demo\",\"COPILOT_ENVIRONMENT_NAME=test
\",\"COPILOT_LB_DNS=demo-Publi-YBFT8XNGALSU-1394143877.us-west-2.elb.amazonaws.com\",\"COPILOT_SERVICE_DISCOVERY_ENDPOINT=test.demo.local\",\"COPILOT_SERVICE_NAME=fron
tend\"
--secrets \"mysecret=/copilot/demo/test/secrets/mysecret\"
--security-groups sg-00e1ea5c957fca925
--app demo --env test
```

> I'm happy to send over a PR for either one.

That'll be awesome! Thank you ❤️

**Comment by codekitchen** [comment]
yeah we'll need to update `--generate-cmd` to have the quote cover both the key and value, but we'll also need to add the escaping backslashes as well. I can push up a PR soon.

**Comment by codekitchen** [comment]
Hi, I started work on implementing this, but I don't think I fully understand #3458 and the fix applied so I'm wary of pushing up a change that may break that fix. I see the fix was to modify `--generate-cmd` to wrap the outputted secret values in double quotes... but those quotes are interpreted by the shell and discarded, so copilot doesn't actually see any difference in the secret value passed in to the `--secrets` flag. How did this actually fix the generation of the task yml file?

Somewhat related: I found in the pflag source where it is parsing these StringToString options, and it is [actually parsing them as csv](https://github.com/spf13/pflag/blob/d5e0c0615acee7028e1e2740a11102313be88de1/string_to_string.go#L85). So a full fix that should work for any arbitrary env var value would be to have `fmtStringMapToString` output csv-escaped fields, and then shell escape that csv output.

**Comment by efekarakus** [comment]
Hi @codekitchen 

> How did this actually fix the generation of the task yml file?

Oooh, I think it was fixing a different issue. The fix in #3458 is for values that contain the `:` character. When a value contains `:`, such as `--env-vars a=arn:` and we try to render it the YAML CloudFormation template it becomes:
```yaml
Environment:
   - Name: a
     Value: arn: # same as Value: {arn: ""}
```
Which causes a YAML malformatted error. 


> those quotes are interpreted by the shell and discarded,

However, I still agree with you even if I do `--env-vars a="arn:"` the quotes get discarded. 

What if we do the following changes:
1. In `fmtStringMapToString` so that we can have multiple inputs to the `--env-vars` and `--secrets` flags:
```go
for _, k := range keys {
	output = append(output, fmt.Sprintf(`\"%s=%v\"`, k, m[k]))
}
```
2.  Update the [CF template](https://github.com/aws/copilot-cli/blob/0e32c1c1f493af7879f3862026fd9987e081bb43/internal/pkg/template/templates/task/cf.yml#L58-L64) for env vars and secrets so that the values are quoted:

```yaml
          Environment:
          {{- range $name, $value := .EnvVars}}
          - Name: {{$name}}
            Value: {{$value | printf "%q"}} # Pipe to printf "%q"
          {{- end}}
          {{- end}}
          {{- if or .SSMParamSecrets .SecretsManagerSecrets}}
          Secrets:
          {{- range $name, $valueFrom := .SSMParamSecrets}}
          - Name: {{$name}}
            ValueFrom: {{$valueFrom | printf "%q"}}
         {{- end}}
```

**Comment by codekitchen** [comment]
Thanks @efekarakus that sounds good.

Using `printf "%q"` to escape in the YAML probably will not work in all cases -- that's escaping/quoting for golang syntax, not yaml syntax. Ideally copilot would do actual yaml escaping of the string there. But I'm not familiar with the options for doing yaml-escaping inside a golang template like that, it could require introducing a new formatting function to the template or something.

I'll work on the first change, let me know your thoughts on the second.

**Comment by efekarakus** [comment]
> golang syntax, not yaml syntax.

Can you give me an example where you think it wouldn't work? 

We do it today for our service container env vars which hasn't caused an issue so far 🤞 :
https://github.com/aws/copilot-cli/blob/850a75e7f69a1467c20b0c833ded9af3d924b404/internal/pkg/template/templates/workloads/partials/cf/envvars-container.yml#L1-L3

I'm worried if we don't do 2. then we still wouldn't have fixed #3458 😅

**Comment by codekitchen** [comment]
I don't have a specific edge case in mind unfortunately, just a general wariness of not escaping for the actual format in play 😁  I suppose it's possible that golang-escaped strings and yaml-escaped strings are equivalent for all possible inputs here, that just seems unlikely to me. But that's good to know that you're already doing the same in defining the service container env vars, it adds confidence that it's unlikely to cause issues in practice.

**Comment by efekarakus** [comment]
This fix is now out in v1.20 🚀 thank you @codekitchen 😍 
Release notes: https://github.com/aws/copilot-cli/releases/tag/v1.20.0 !
Blog post: https://aws.github.io/copilot-cli/blogs/release-v120/
